dds/DCPS/XTypes/DynamicData.cpp

```diff
@@ -2193,6 +2302,416 @@ const char* DynamicData::typekind_to_string(TypeKind tk) const
+      }
+      return false;
+    }
+    std::stringstream os;
+    char_helper<ACE_CDR::WChar>(os, my_wchar);
+    type_string += " = L'" + os.str() + "'\n";
+    break;
+  }
+  }
+  return true;
+}
+
+bool print_dynamic_data(DynamicData& dd, DCPS::String& type_string, DCPS::String& indent)
+{
+  DCPS::String member_name;
+  DCPS::String type_name;
+  DynamicData temp_dd;
+  switch (dd.type()->get_kind()) {
+  case TK_INT8:
+  case TK_UINT8:
+  case TK_INT16:
+  case TK_UINT16:
+  case TK_ENUM:
+  case TK_INT32:
+  case TK_UINT32:
+  case TK_INT64:
+  case TK_UINT64:
+  case TK_BOOLEAN:
+  case TK_BYTE:
+  case TK_CHAR8:
+  case TK_CHAR16:
+    if (!print_integral_value(dd, type_string, dd.type())) {
+      return false;
+    }
+    break;
+  case TK_FLOAT32: {
+    ACE_CDR::Float my_float;
+    if (dd.get_float32_value(my_float, 0) != DDS::RETCODE_OK) {
+      if (DCPS::log_level >= DCPS::LogLevel::Notice) {
+        ACE_ERROR((LM_NOTICE, "(%P|%t) NOTICE: print_dynamic_data: failed to get_float32_value\n"));
+      }
+      return false;
+    }
+    std::stringstream os;
+    os << my_float;
+    type_string += " = " + os.str() + "\n";
+    break;
+  }
+  case TK_FLOAT64: {
+    ACE_CDR::Double my_double;
+    if (dd.get_float64_value(my_double, 0) != DDS::RETCODE_OK) {
+      if (DCPS::log_level >= DCPS::LogLevel::Notice) {
+        ACE_ERROR((LM_NOTICE, "(%P|%t) NOTICE: print_dynamic_data: failed to get_float64_value\n"));
+      }
+      return false;
+    }
+    std::stringstream os;
+    os << my_double;
+    type_string += " = " + os.str() + "\n";
+    break;
+  }
+  case TK_FLOAT128: {
+    ACE_CDR::LongDouble my_longdouble;
+    if (dd.get_float128_value(my_longdouble, 0) != DDS::RETCODE_OK) {
+      if (DCPS::log_level >= DCPS::LogLevel::Notice) {
+        ACE_ERROR((LM_NOTICE, "(%P|%t) NOTICE: print_dynamic_data: failed to get_float128_value\n"));
+      }
+      return false;
+    }
+    std::stringstream os;
+    os << my_longdouble;
+    type_string += " = " + os.str() + "\n";
+    break;
+  }
+  case TK_STRING8: {
+    ACE_CDR::Char* my_string = 0;
+    if (dd.get_string_value(my_string, 0) != DDS::RETCODE_OK) {
+      if (DCPS::log_level >= DCPS::LogLevel::Notice) {
+        ACE_ERROR((LM_NOTICE, "(%P|%t) NOTICE: print_dynamic_data: failed to get_string_value\n"));
+      }
+      return false;
+    }
+    type_string += DCPS::String(" = \"") + my_string + "\"\n";
+    break;
+  }
+  case TK_STRING16: {
+    ACE_CDR::WChar* my_wstring = 0;
+    if (dd.get_wstring_value(my_wstring, 0) != DDS::RETCODE_OK) {
+      if (DCPS::log_level >= DCPS::LogLevel::Notice) {
+        ACE_ERROR((LM_NOTICE, "(%P|%t) NOTICE: print_dynamic_data: failed to get_wstring_value\n"));
+      }
+      return false;
+    }
+    std::stringstream os;
+    for (size_t i = 0; my_wstring[i] != 0; ++i) {
+      char_helper<ACE_CDR::WChar>(os, my_wstring[i]);
+    }
+    type_string += " = L\"" + os.str() + "\"\n";
```

Comment:
Turn this into a `string_helper` template function and also apply it to narrow strings.